### DISCUSSION TITLE
Extension attributes for span groups

### DISCUSSION THREAD

**Discussion by sztal** [explosion/spaCy#13036] [New Features & Project Ideas]
Hi!

First, thanks for all the great work at Explosion! I wonder whether there are any plans for implementing the extension attribute mechanism for span groups? 

Span groups seem to be opening some very interesting possibilities for modeling text at the level between tokens and sentences, so being able to enhance them with additional methods would be super useful.

**Comment by adrianeboyd** [comment]
Thanks for the feedback!

Do you mean that each span group would have its own set of custom extensions that only apply to its own spans? Or if that's not what you meant, could you give an example of what this might look like in practice?

**Reply by sztal** [reply]
Well, not exactly. I rather thought about treating them in the same way as `Token`, `Span` and `Doc` objects, so there would be methods like `SpanGroup.set_extension` etc.

Here is a concrete, even if naive, example.

```python
import spacy
from spacy.tokens import SpanGroup
from more_itertools import unique_everseen

nlp = spacy.load("en_core_web_trf")
doc = nlp("The sudden arrival.")
nominals = [[doc[:1], doc[-2:-1]], [doc[1:2]]]

for i, nominal in enumerate(nominals):
    doc.spans[f"nominal_{i}"] = nominal

def tokens(spans):
    yield from unique_everseen((tok for span in spans for tok in span), key=lambda tok: tok.i)

spans_tokens = [ [*tokens(spans)] for spans in doc.spans.values() ]

# NEW FEATURE
SpanGroup.set_extension("tokens", getter=tokens)
assert spans_tokens == [ [*span._.tokens] for span in doc.spans.values() ]
```

**Reply by adrianeboyd** [reply]
Thanks for the example! The custom extension data is saved internally based only on the position in the doc, so custom `SpanGroup` extensions would need to be implemented in a separate way (possibly it could be more like some of the other custom user hooks?), but it's an interesting suggestion...

**Reply by sztal** [reply]
I see, user hooks would also be great, although, to the extent I understand this mechanism, it would less flexible as there is only a small set of pre-defined user hooks one can define (vector hooks, similarity hooks etc.), right?

**Comment by svlandeg** [comment]
Hi @sztal! This is an interesting feature request. We might not have time to look at this right now, but I'm moving this idea to the Issue Tracker to better keep track of it for the future.
